velox/exec/tests/RowContainerTest.cpp

```diff
@@ -466,29 +465,79 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
     }
   }
 
-  template <typename T, typename V>
-  void testOrderAndNullsFirstVariations(
+  template <bool canHandleNulls>
+  void testRowContainerEqualsAPI(
+      const TypePtr& type,
+      const VectorPtr& lhs,
+      const VectorPtr& rhs,
+      const VectorPtr& expected) {
+    EXPECT_EQ(lhs->size(), rhs->size());
+    EXPECT_EQ(lhs->size(), expected->size());
+
+    auto numRows = lhs->size();
+    auto rowContainer = makeRowContainer({type}, {type}, false);
+    DecodedVector lhsDecoded(*lhs);
+    auto rows = storeRows(lhsDecoded, lhs->size(), *rowContainer);
+
+    DecodedVector rhsDecoded(*rhs);
+
+    int32_t index{0};
+    for (auto row : rows) {
+      ASSERT_EQ(
+          expected->asFlatVector<bool>()->valueAt(index),
+          rowContainer->equals<canHandleNulls>(
```

Comment:
We are always creating an rhs of a primitive type ([L606](https://github.com/facebookincubator/velox/blob/74c7f49418bb800b45b8bd6b9a51590cd1985264/velox/exec/tests/RowContainerTest.cpp#L606)) regardless of whether lhs is primitive or complex(array/map/row). Then we end up comparing rowContainer->equals() which dynamically dispatches based on the rhsDecoded's type which will always be a primitive.

Should we instead make sure to create the RHS vector of the same type as LHS?

Reply:
That's a good point.